File: swift/BCLifeHash/Sources/BCLifeHash/Gradients.swift
func grayscale() -> ColorFunc {
    blend2(.black, .white)
}

func selectGrayscale(_ entropy: BitEnumerator) -> ColorFunc {
    entropy.next() ? grayscale() : reverse(grayscale())
}

func makeHue(_ t: Double) -> Color {
    HSBColor.fromHue(t).color()
}

func spectrum() -> ColorFunc {
    blend([
        Color.fromUInt8Values(0, 168, 222),
        Color.fromUInt8Values(51, 51, 145),
        Color.fromUInt8Values(233, 19, 136),
        Color.fromUInt8Values(235, 45, 46),
        Color.fromUInt8Values(253, 233, 43),
        Color.fromUInt8Values(0, 158, 84),
        Color.fromUInt8Values(0, 168, 222),
    ])
}

func spectrumCMYKSafe() -> ColorFunc {
    blend([
        Color.fromUInt8Values(0, 168, 222),
        Color.fromUInt8Values(41, 60, 130),
        Color.fromUInt8Values(210, 59, 130),
        Color.fromUInt8Values(217, 63, 53),
        Color.fromUInt8Values(244, 228, 81),
        Color.fromUInt8Values(0, 158, 84),
        Color.fromUInt8Values(0, 168, 222),
    ])
}

func adjustForLuminance(_ color: Color, _ contrastColor: Color) -> Color {
    let lum = color.luminance()
    let contrastLum = contrastColor.luminance()
    let threshold = 0.6
    let offset = abs(lum - contrastLum)
    if offset > threshold {
        return color
    }
    let boost = 0.7
    let t = lerp(0.0, threshold, boost, 0.0, offset)
    if contrastLum > lum {
        return color.darken(t).burn(t * 0.6)
    }
    return color.lighten(t).burn(t * 0.6)
}

func monochromatic(_ entropy: BitEnumerator, _ hueGenerator: @escaping ColorFunc) -> ColorFunc {
    let hue = entropy.nextFrac()
    let isTint = entropy.next()
    let isReversed = entropy.next()
    let keyAdvance = entropy.nextFrac() * 0.3 + 0.05
    let neutralAdvance = entropy.nextFrac() * 0.3 + 0.05

    var keyColor = hueGenerator(hue)

    let contrastBrightness: Double
    if isTint {
        contrastBrightness = 1.0
        keyColor = keyColor.darken(0.5)
    } else {
        contrastBrightness = 0.0
    }

    let gs = grayscale()
    let neutralColor = gs(contrastBrightness)

    let keyColor2 = keyColor.lerpTo(neutralColor, keyAdvance)
    let neutralColor2 = neutralColor.lerpTo(keyColor, neutralAdvance)

    let gradient = blend2(keyColor2, neutralColor2)
    return isReversed ? reverse(gradient) : gradient
}

func monochromaticFiducial(_ entropy: BitEnumerator) -> ColorFunc {
    let hue = entropy.nextFrac()
    let isReversed = entropy.next()
    let isTint = entropy.next()

    let contrastColor = isTint ? Color.white : Color.black
    let spec = spectrumCMYKSafe()
    let keyColor = adjustForLuminance(spec(hue), contrastColor)

    let gradient = blend([keyColor, contrastColor, keyColor])
    return isReversed ? reverse(gradient) : gradient
}

func complementary(_ entropy: BitEnumerator, _ hueGenerator: @escaping ColorFunc) -> ColorFunc {
    let spectrum1 = entropy.nextFrac()
    let spectrum2 = modulo(spectrum1 + 0.5, 1.0)
    let lighterAdvance = entropy.nextFrac() * 0.3
    let darkerAdvance = entropy.nextFrac() * 0.3
    let isReversed = entropy.next()

    let color1 = hueGenerator(spectrum1)
    let color2 = hueGenerator(spectrum2)

    let luma1 = color1.luminance()
    let luma2 = color2.luminance()

    let darkerColor: Color
    let lighterColor: Color
    if luma1 > luma2 {
        darkerColor = color2
        lighterColor = color1
    } else {
        darkerColor = color1
        lighterColor = color2
    }

    let adjustedLighter = lighterColor.lighten(lighterAdvance)
    let adjustedDarker = darkerColor.darken(darkerAdvance)

    let gradient = blend2(adjustedDarker, adjustedLighter)
    return isReversed ? reverse(gradient) : gradient
}

func complementaryFiducial(_ entropy: BitEnumerator) -> ColorFunc {
    let spectrum1 = entropy.nextFrac()
    let spectrum2 = modulo(spectrum1 + 0.5, 1.0)
    let isTint = entropy.next()
    let isReversed = entropy.next()
    let neutralColorBias = entropy.next()

    let neutralColor = isTint ? Color.white : Color.black
    let spec = spectrumCMYKSafe()
    let color1 = spec(spectrum1)
    let color2 = spec(spectrum2)

    let biasColor = neutralColorBias ? color1 : color2
    let biasedNeutralColor = neutralColor.lerpTo(biasColor, 0.2).burn(0.1)

    let gradient = blend([
        adjustForLuminance(color1, biasedNeutralColor),
        biasedNeutralColor,
        adjustForLuminance(color2, biasedNeutralColor),
    ])
    return isReversed ? reverse(gradient) : gradient
}

func triadic(_ entropy: BitEnumerator, _ hueGenerator: @escaping ColorFunc) -> ColorFunc {
    let spectrum1 = entropy.nextFrac()
    let spectrum2 = modulo(spectrum1 + 1.0 / 3.0, 1.0)
    let spectrum3 = modulo(spectrum1 + 2.0 / 3.0, 1.0)
    let lighterAdvance = entropy.nextFrac() * 0.3
    let darkerAdvance = entropy.nextFrac() * 0.3
    let isReversed = entropy.next()

    let color1 = hueGenerator(spectrum1)
    let color2 = hueGenerator(spectrum2)
    let color3 = hueGenerator(spectrum3)

    var colors = [color1, color2, color3]
    colors.sort { $0.luminance() < $1.luminance() }

    let darkerColor = colors[0]
    let middleColor = colors[1]
    let lighterColor = colors[2]

    let adjustedLighter = lighterColor.lighten(lighterAdvance)
    let adjustedDarker = darkerColor.darken(darkerAdvance)

    let gradient = blend([adjustedLighter, middleColor, adjustedDarker])
    return isReversed ? reverse(gradient) : gradient
}

func triadicFiducial(_ entropy: BitEnumerator) -> ColorFunc {
    let spectrum1 = entropy.nextFrac()
    let spectrum2 = modulo(spectrum1 + 1.0 / 3.0, 1.0)
    let spectrum3 = modulo(spectrum1 + 2.0 / 3.0, 1.0)
    let isTint = entropy.next()
    let neutralInsertIndex = Int(entropy.nextUInt8() % 2 + 1)
    let isReversed = entropy.next()

    let neutralColor = isTint ? Color.white : Color.black

    let spec = spectrumCMYKSafe()
    var colors = [spec(spectrum1), spec(spectrum2), spec(spectrum3)]

    switch neutralInsertIndex {
    case 1:
        colors[0] = adjustForLuminance(colors[0], neutralColor)
        colors[1] = adjustForLuminance(colors[1], neutralColor)
        colors[2] = adjustForLuminance(colors[2], colors[1])
    case 2:
        colors[1] = adjustForLuminance(colors[1], neutralColor)
        colors[2] = adjustForLuminance(colors[2], neutralColor)
        colors[0] = adjustForLuminance(colors[0], colors[1])
    default:
        preconditionFailure("Internal error")
    }

    colors.insert(neutralColor, at: neutralInsertIndex)

    let gradient = blend(colors)
    return isReversed ? reverse(gradient) : gradient
}

func analogous(_ entropy: BitEnumerator, _ hueGenerator: @escaping ColorFunc) -> ColorFunc {
    let spectrum1 = entropy.nextFrac()
    let spectrum2 = modulo(spectrum1 + 1.0 / 12.0, 1.0)
    let spectrum3 = modulo(spectrum1 + 2.0 / 12.0, 1.0)
    let spectrum4 = modulo(spectrum1 + 3.0 / 12.0, 1.0)
    let advance = entropy.nextFrac() * 0.5 + 0.2
    let isReversed = entropy.next()

    let color1 = hueGenerator(spectrum1)
    let color2 = hueGenerator(spectrum2)
    let color3 = hueGenerator(spectrum3)
    let color4 = hueGenerator(spectrum4)

    let darkestColor: Color
    let darkColor: Color
    let lightColor: Color
    let lightestColor: Color

    if color1.luminance() < color4.luminance() {
        darkestColor = color1
        darkColor = color2
        lightColor = color3
        lightestColor = color4
    } else {
        darkestColor = color4
        darkColor = color3
        lightColor = color2
        lightestColor = color1
    }

    let adjustedDarkest = darkestColor.darken(advance)
    let adjustedDark = darkColor.darken(advance / 2.0)
    let adjustedLight = lightColor.lighten(advance / 2.0)
    let adjustedLightest = lightestColor.lighten(advance)

    let gradient = blend([adjustedDarkest, adjustedDark, adjustedLight, adjustedLightest])
    return isReversed ? reverse(gradient) : gradient
}

func analogousFiducial(_ entropy: BitEnumerator) -> ColorFunc {
    let spectrum1 = entropy.nextFrac()
    let spectrum2 = modulo(spectrum1 + 1.0 / 10.0, 1.0)
    let spectrum3 = modulo(spectrum1 + 2.0 / 10.0, 1.0)
    let isTint = entropy.next()
    let neutralInsertIndex = Int(entropy.nextUInt8() % 2 + 1)
    let isReversed = entropy.next()

    let neutralColor = isTint ? Color.white : Color.black

    let spec = spectrumCMYKSafe()
    var colors = [spec(spectrum1), spec(spectrum2), spec(spectrum3)]

    switch neutralInsertIndex {
    case 1:
        colors[0] = adjustForLuminance(colors[0], neutralColor)
        colors[1] = adjustForLuminance(colors[1], neutralColor)
        colors[2] = adjustForLuminance(colors[2], colors[1])
    case 2:
        colors[1] = adjustForLuminance(colors[1], neutralColor)
        colors[2] = adjustForLuminance(colors[2], neutralColor)
        colors[0] = adjustForLuminance(colors[0], colors[1])
    default:
        preconditionFailure("Internal error")
    }

    colors.insert(neutralColor, at: neutralInsertIndex)

    let gradient = blend(colors)
    return isReversed ? reverse(gradient) : gradient
}

func selectGradient(_ entropy: BitEnumerator, _ version: Version) -> ColorFunc {
    if version == .grayscaleFiducial {
        return selectGrayscale(entropy)
    }

    let value = entropy.nextUInt2()

    switch value {
    case 0:
        switch version {
        case .version1:
            return monochromatic(entropy, makeHue)
        case .version2, .detailed:
            let spec = spectrumCMYKSafe()
            return monochromatic(entropy, spec)
        case .fiducial:
            return monochromaticFiducial(entropy)
        case .grayscaleFiducial:
            return grayscale()
        }
    case 1:
        switch version {
        case .version1:
            let spec = spectrum()
            return complementary(entropy, spec)
        case .version2, .detailed:
            let spec = spectrumCMYKSafe()
            return complementary(entropy, spec)
        case .fiducial:
            return complementaryFiducial(entropy)
        case .grayscaleFiducial:
            return grayscale()
        }
    case 2:
        switch version {
        case .version1:
            let spec = spectrum()
            return triadic(entropy, spec)
        case .version2, .detailed:
            let spec = spectrumCMYKSafe()
            return triadic(entropy, spec)
        case .fiducial:
            return triadicFiducial(entropy)
        case .grayscaleFiducial:
            return grayscale()
        }
    case 3:
        switch version {
        case .version1:
            let spec = spectrum()
            return analogous(entropy, spec)
        case .version2, .detailed:
            let spec = spectrumCMYKSafe()
            return analogous(entropy, spec)
        case .fiducial:
            return analogousFiducial(entropy)
        case .grayscaleFiducial:
            return grayscale()
        }
    default:
        return grayscale()
    }
}

File: python/bc-lifehash/src/bc_lifehash/_gradients.py
from __future__ import annotations

from typing import TYPE_CHECKING

from ._bit_enumerator import BitEnumerator
from ._color import BLACK, WHITE, Color, lerp, modulo
from ._color_func import ColorFunc, blend, blend2, reverse
from ._hsb_color import HSBColor

if TYPE_CHECKING:
    from ._lifehash import Version


def _grayscale() -> ColorFunc:
    return blend2(BLACK, WHITE)


def _select_grayscale(entropy: BitEnumerator) -> ColorFunc:
    if entropy.next():
        return _grayscale()
    return reverse(_grayscale())


def _make_hue(t: float) -> Color:
    return HSBColor.from_hue(t).color()


def _spectrum() -> ColorFunc:
    return blend([
        Color.from_uint8_values(0, 168, 222),
        Color.from_uint8_values(51, 51, 145),
        Color.from_uint8_values(233, 19, 136),
        Color.from_uint8_values(235, 45, 46),
        Color.from_uint8_values(253, 233, 43),
        Color.from_uint8_values(0, 158, 84),
        Color.from_uint8_values(0, 168, 222),
    ])


def _spectrum_cmyk_safe() -> ColorFunc:
    return blend([
        Color.from_uint8_values(0, 168, 222),
        Color.from_uint8_values(41, 60, 130),
        Color.from_uint8_values(210, 59, 130),
        Color.from_uint8_values(217, 63, 53),
        Color.from_uint8_values(244, 228, 81),
        Color.from_uint8_values(0, 158, 84),
        Color.from_uint8_values(0, 168, 222),
    ])


def _adjust_for_luminance(color: Color, contrast_color: Color) -> Color:
    lum = color.luminance()
    contrast_lum = contrast_color.luminance()
    threshold = 0.6
    offset = abs(lum - contrast_lum)
    if offset > threshold:
        return color
    boost = 0.7
    t = lerp(0.0, threshold, boost, 0.0, offset)
    if contrast_lum > lum:
        return color.darken(t).burn(t * 0.6)
    return color.lighten(t).burn(t * 0.6)


def _monochromatic(entropy: BitEnumerator, hue_generator: ColorFunc) -> ColorFunc:
    hue = entropy.next_frac()
    is_tint = entropy.next()
    is_reversed = entropy.next()
    key_advance = entropy.next_frac() * 0.3 + 0.05
    neutral_advance = entropy.next_frac() * 0.3 + 0.05

    key_color = hue_generator(hue)

    if is_tint:
        contrast_brightness = 1.0
        key_color = key_color.darken(0.5)
    else:
        contrast_brightness = 0.0

    gs = _grayscale()
    neutral_color = gs(contrast_brightness)

    key_color_2 = key_color.lerp_to(neutral_color, key_advance)
    neutral_color_2 = neutral_color.lerp_to(key_color, neutral_advance)

    gradient = blend2(key_color_2, neutral_color_2)
    return reverse(gradient) if is_reversed else gradient


def _monochromatic_fiducial(entropy: BitEnumerator) -> ColorFunc:
    hue = entropy.next_frac()
    is_reversed = entropy.next()
    is_tint = entropy.next()

    contrast_color = WHITE if is_tint else BLACK
    spec = _spectrum_cmyk_safe()
    key_color = _adjust_for_luminance(spec(hue), contrast_color)

    gradient = blend([key_color, contrast_color, key_color])
    return reverse(gradient) if is_reversed else gradient


def _complementary(entropy: BitEnumerator, hue_generator: ColorFunc) -> ColorFunc:
    spectrum1 = entropy.next_frac()
    spectrum2 = modulo(spectrum1 + 0.5, 1.0)
    lighter_advance = entropy.next_frac() * 0.3
    darker_advance = entropy.next_frac() * 0.3
    is_reversed = entropy.next()

    color1 = hue_generator(spectrum1)
    color2 = hue_generator(spectrum2)

    luma1 = color1.luminance()
    luma2 = color2.luminance()

    if luma1 > luma2:
        darker_color, lighter_color = color2, color1
    else:
        darker_color, lighter_color = color1, color2

    adjusted_lighter = lighter_color.lighten(lighter_advance)
    adjusted_darker = darker_color.darken(darker_advance)

    gradient = blend2(adjusted_darker, adjusted_lighter)
    return reverse(gradient) if is_reversed else gradient


def _complementary_fiducial(entropy: BitEnumerator) -> ColorFunc:
    spectrum1 = entropy.next_frac()
    spectrum2 = modulo(spectrum1 + 0.5, 1.0)
    is_tint = entropy.next()
    is_reversed = entropy.next()
    neutral_color_bias = entropy.next()

    neutral_color = WHITE if is_tint else BLACK
    spec = _spectrum_cmyk_safe()
    color1 = spec(spectrum1)
    color2 = spec(spectrum2)

    bias_color = color1 if neutral_color_bias else color2
    biased_neutral_color = neutral_color.lerp_to(bias_color, 0.2).burn(0.1)

    gradient = blend([
        _adjust_for_luminance(color1, biased_neutral_color),
        biased_neutral_color,
        _adjust_for_luminance(color2, biased_neutral_color),
    ])
    return reverse(gradient) if is_reversed else gradient


def _triadic(entropy: BitEnumerator, hue_generator: ColorFunc) -> ColorFunc:
    spectrum1 = entropy.next_frac()
    spectrum2 = modulo(spectrum1 + 1.0 / 3.0, 1.0)
    spectrum3 = modulo(spectrum1 + 2.0 / 3.0, 1.0)
    lighter_advance = entropy.next_frac() * 0.3
    darker_advance = entropy.next_frac() * 0.3
    is_reversed = entropy.next()

    color1 = hue_generator(spectrum1)
    color2 = hue_generator(spectrum2)
    color3 = hue_generator(spectrum3)

    colors = sorted([color1, color2, color3], key=lambda c: c.luminance())
    darker_color = colors[0]
    middle_color = colors[1]
    lighter_color = colors[2]

    adjusted_lighter = lighter_color.lighten(lighter_advance)
    adjusted_darker = darker_color.darken(darker_advance)

    gradient = blend([adjusted_lighter, middle_color, adjusted_darker])
    return reverse(gradient) if is_reversed else gradient


def _triadic_fiducial(entropy: BitEnumerator) -> ColorFunc:
    spectrum1 = entropy.next_frac()
    spectrum2 = modulo(spectrum1 + 1.0 / 3.0, 1.0)
    spectrum3 = modulo(spectrum1 + 2.0 / 3.0, 1.0)
    is_tint = entropy.next()
    neutral_insert_index = (entropy.next_uint8() % 2 + 1)
    is_reversed = entropy.next()

    neutral_color = WHITE if is_tint else BLACK

    spec = _spectrum_cmyk_safe()
    colors = [spec(spectrum1), spec(spectrum2), spec(spectrum3)]

    if neutral_insert_index == 1:
        colors[0] = _adjust_for_luminance(colors[0], neutral_color)
        colors[1] = _adjust_for_luminance(colors[1], neutral_color)
        colors[2] = _adjust_for_luminance(colors[2], colors[1])
    elif neutral_insert_index == 2:
        colors[1] = _adjust_for_luminance(colors[1], neutral_color)
        colors[2] = _adjust_for_luminance(colors[2], neutral_color)
        colors[0] = _adjust_for_luminance(colors[0], colors[1])
    else:
        raise RuntimeError("Internal error")

    colors.insert(neutral_insert_index, neutral_color)

    gradient = blend(colors)
    return reverse(gradient) if is_reversed else gradient


def _analogous(entropy: BitEnumerator, hue_generator: ColorFunc) -> ColorFunc:
    spectrum1 = entropy.next_frac()
    spectrum2 = modulo(spectrum1 + 1.0 / 12.0, 1.0)
    spectrum3 = modulo(spectrum1 + 2.0 / 12.0, 1.0)
    spectrum4 = modulo(spectrum1 + 3.0 / 12.0, 1.0)
    advance = entropy.next_frac() * 0.5 + 0.2
    is_reversed = entropy.next()

    color1 = hue_generator(spectrum1)
    color2 = hue_generator(spectrum2)
    color3 = hue_generator(spectrum3)
    color4 = hue_generator(spectrum4)

    if color1.luminance() < color4.luminance():
        darkest_color, dark_color, light_color, lightest_color = (
            color1, color2, color3, color4
        )
    else:
        darkest_color, dark_color, light_color, lightest_color = (
            color4, color3, color2, color1
        )

    adjusted_darkest = darkest_color.darken(advance)
    adjusted_dark = dark_color.darken(advance / 2.0)
    adjusted_light = light_color.lighten(advance / 2.0)
    adjusted_lightest = lightest_color.lighten(advance)

    gradient = blend([
        adjusted_darkest, adjusted_dark, adjusted_light, adjusted_lightest
    ])
    return reverse(gradient) if is_reversed else gradient


def _analogous_fiducial(entropy: BitEnumerator) -> ColorFunc:
    spectrum1 = entropy.next_frac()
    spectrum2 = modulo(spectrum1 + 1.0 / 10.0, 1.0)
    spectrum3 = modulo(spectrum1 + 2.0 / 10.0, 1.0)
    is_tint = entropy.next()
    neutral_insert_index = (entropy.next_uint8() % 2 + 1)
    is_reversed = entropy.next()

    neutral_color = WHITE if is_tint else BLACK

    spec = _spectrum_cmyk_safe()
    colors = [spec(spectrum1), spec(spectrum2), spec(spectrum3)]

    if neutral_insert_index == 1:
        colors[0] = _adjust_for_luminance(colors[0], neutral_color)
        colors[1] = _adjust_for_luminance(colors[1], neutral_color)
        colors[2] = _adjust_for_luminance(colors[2], colors[1])
    elif neutral_insert_index == 2:
        colors[1] = _adjust_for_luminance(colors[1], neutral_color)
        colors[2] = _adjust_for_luminance(colors[2], neutral_color)
        colors[0] = _adjust_for_luminance(colors[0], colors[1])
    else:
        raise RuntimeError("Internal error")

    colors.insert(neutral_insert_index, neutral_color)

    gradient = blend(colors)
    return reverse(gradient) if is_reversed else gradient


def select_gradient(entropy: BitEnumerator, version: Version) -> ColorFunc:
    from ._lifehash import Version

    if version == Version.GRAYSCALE_FIDUCIAL:
        return _select_grayscale(entropy)

    value = entropy.next_uint2()

    if value == 0:
        if version == Version.VERSION1:
            return _monochromatic(entropy, _make_hue)
        if version in (Version.VERSION2, Version.DETAILED):
            return _monochromatic(entropy, _spectrum_cmyk_safe())
        if version == Version.FIDUCIAL:
            return _monochromatic_fiducial(entropy)
        return _grayscale()
    elif value == 1:
        if version == Version.VERSION1:
            return _complementary(entropy, _spectrum())
        if version in (Version.VERSION2, Version.DETAILED):
            return _complementary(entropy, _spectrum_cmyk_safe())
        if version == Version.FIDUCIAL:
            return _complementary_fiducial(entropy)
        return _grayscale()
    elif value == 2:
        if version == Version.VERSION1:
            return _triadic(entropy, _spectrum())
        if version in (Version.VERSION2, Version.DETAILED):
            return _triadic(entropy, _spectrum_cmyk_safe())
        if version == Version.FIDUCIAL:
            return _triadic_fiducial(entropy)
        return _grayscale()
    elif value == 3:
        if version == Version.VERSION1:
            return _analogous(entropy, _spectrum())
        if version in (Version.VERSION2, Version.DETAILED):
            return _analogous(entropy, _spectrum_cmyk_safe())
        if version == Version.FIDUCIAL:
            return _analogous_fiducial(entropy)
        return _grayscale()
    else:
        return _grayscale()
